[FLINK-18045] Fix Kerberos credentials checking

krasinski opened the pull request:
## What is the purpose of the change

This pull request brings back checking if the enabled security method is Kerberos, that's an old bug reintroduced during some refactoring.
Code assuming that the enabled security method is Kerberos caused problems on MapR secured clusters when deploying on YARN.

## Brief change log

  - Checking if the auth method is Kerberos was added to `org.apache.flink.runtime.util.HadoopUtils.isCredentialsConfigured` method.

## Verifying this change
This change added tests and can be verified as follows:

  - Added a test that validates that the modified method doesn't check for Kerberos credentials when auth method is other than Kerberos

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

walterddr commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/util/HadoopUtilsTest.java:
```diff
@@ -0,0 +1,49 @@
+public class HadoopUtilsTest {
+
+	@Test
+	public void testShouldNotCheckKerberosCredentialsForOtherAuthMethods() throws Exception {
```
can you also add another test that verifies Kerberos-based auth actually walk into the if branch?

krasinski replied:
done, also more tests added

walterddr commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/util/HadoopUtilsTest.java:
```diff
@@ -0,0 +1,49 @@
+		assertTrue(result);
+	}
+
+	private static Configuration getHadoopConfigWithAuthMethodOtherThanKerberos() {
```
we can probably generalize this so we can test both side 
```
Configuration getHadoopConfigWithAutMethod(AuthenticationMethod method) {
  // ...
```

krasinski replied:
done

walterddr commented on flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/util/HadoopUtils.java:
```diff
@@ -113,7 +113,8 @@ public static Configuration getHadoopConfiguration(org.apache.flink.configuratio
 	public static boolean isCredentialsConfigured(UserGroupInformation ugi, boolean useTicketCache) throws Exception {
-		if (UserGroupInformation.isSecurityEnabled()) {
+		if (UserGroupInformation.isSecurityEnabled()
+			&& ugi.getAuthenticationMethod() == UserGroupInformation.AuthenticationMethod.KERBEROS) {
```
adding this line means the function name should not be `isCredentialsConfigured` but more like `isKerberosCredentialsConfigured`. I was wondering if we should change the public API naming.

krasinski replied:
done, please let me know if you decide that changing that public method signature can be a problem - I'll bring back the old method name

tillrohrmann replied:
Hmm, I think the current method name is a bit misleading because it also returns `true` if no security has been enabled. In the current form, this method should rather be called `areKerberosCredentialsConfiguredProperlyIfKerberosAuthenticationModeIsEnabled`. This is quite a mouthful and an indicator that this method is too specific.

What about creating a `isKerberosSecurityEnabled` method and also having a `isKerberosCredentialsValid` which we only call if `isKerberosSecurityEnabled` returns true?

walterddr replied:
+1 on this. 
also do you think we should keep the `isCredentialsConfigured` API available for 1 more release and it can simply be:
```
@Deprecated
public static boolean isCredentialsConfigured(UserGroupInformation ugi, boolean useTicketCache) throws Exception {
    return isKerberosSecurityEnabled(ugi) && isKerberosCredentialsValid(ugi, useTicketCache);
}
```

krasinski commented:
thank you for the review @walterddr, I've addressed all the comments and added some more unit tests

I've also introduced a small change in `hasHDFSDelegationToken` method - taking advantage of the UGI parameter instead of retrieving it one more time seems more natural

walterddr commented:
thanks for the quick update. in general the patch looks good to me. I still have only one concern regarding the alternation of 2 public APIs. but IMO `HadoopUtils` are more of an internal runtime util that was shared between different deployment environments. thus I am not sure whether the standard public API alternation rules applies here. 

I would wait for @tillrohrmann for a final review.

walterddr commented on flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/util/HadoopUtils.java:
```diff
@@ -112,11 +112,12 @@ public static Configuration getHadoopConfiguration(org.apache.flink.configuratio
 			if (useTicketCache && !ugi.hasKerberosCredentials()) {
 				// a delegation token is an adequate substitute in most cases
-				if (!HadoopUtils.hasHDFSDelegationToken()) {
+				if (!HadoopUtils.hasHDFSDelegationToken(ugi)) {
```
This change actually causes the problem: The `ugi` passed into the Kerberos credential check is not necessarily the once for delegation token check.

Specifically in Flink, credentials are checked against the `loginUser` and the delegation token always check against `currentUser` in the case of a delegation or proxy user impersonation. 
See: [HadoopModule.java#L137](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/security/modules/HadoopModule.java#L137)

I think this is the reason why the CI test fails @tillrohrmann but I can't be sure. @krasinski can you revert this piece of change and try again?

tillrohrmann commented:
I think I know what the problem is: We are reusing forks and another test has loaded `sun.security.krb5.Config` before without having specified the `java.security.krb5.realm` and `java.security.krb5.kdc`. That's why `sun.security.krb5.Config.getDefaultRealm` fails. 

@krasinski could you try to add `sun.security.krb5.Config.refresh();` as the last line to `HadoopUtilsTest.setPropertiesToEnableKerberosConfigInit`? This should try to reload the Kerberos configuration.

tillrohrmann commented:
I've pushed a small hotfix addressing my comments. @krasinski and @walterddr could you check whether the changes (especially splitting up the method in `HadoopUtils`) make sense to you?

krasinski commented:
Looks good to me.
One additional thing comes to my mind: when I started working on the fix to address review comments I've encountered a problem with java modules when using `sun.security.krb5.Config.refresh()` so adding a new line in java 11 profile compiler args in the main `pom.xml` might be needed:
`--add-exports=java.security.jgss/sun.security.krb5=ALL-UNNAMED`

tillrohrmann commented:
@krasinski thanks for the pointer. I will update the PR accordingly.

walterddr commented:
Thanks for the fix @tillrohrmann . Looks good to me :-)
